compiler/passes/denormalize.cpp

```diff
@@ -0,0 +1,602 @@
+#include "baseAST.h"
+#include "stlUtil.h"
+#include "expr.h"
+#include "stmt.h"
+#include "symbol.h"
+#include "type.h"
+#include "LoopStmt.h"
+#include "CForLoop.h"
+#include "WhileStmt.h"
+
+//helper datastructures/types
+
+//maybe more memoization than caching
+std::map<Expr*,bool> safeExprCache;
+std::map<FnSymbol*,bool> globalManipFuncCache;
+
+typedef std::map<SymExpr*, std::pair<Expr*,Type*> > ActualUseDefCastMap;
+
+//prototypes
+bool isExprSafeForReorder(Expr * e);
+bool canPrimMoveCreateCommunication(CallExpr* ce);
+inline bool possibleDepInBetween(Expr* e1, Expr* e2);
+inline bool requiresCast(Type* t);
+inline bool isIntegerPromotionPrimitive(PrimitiveTag tag);
+inline bool isNonEssentialPrimitive(CallExpr* ce);
+bool isSafePrimitive(CallExpr* ce);
+
+bool isDenormalizable(Symbol* sym,
+    Map<Symbol*,Vec<SymExpr*>*> defMap,
+    Map<Symbol*,Vec<SymExpr*>*> useMap, SymExpr** useOut, Expr** defOut,
+    Type** castTo);
+
+void denormalizeActuals(CallExpr* ce,
+    Map<Symbol*,Vec<SymExpr*>*> defMap,
+    Map<Symbol*,Vec<SymExpr*>*> useMap,
+    ActualUseDefCastMap actualUseDefMap);
+
+void denormalize(void);
+void denormalize(FnSymbol *fn);
+void denormalize(Expr* def, SymExpr* use, Type* castTo);
+
+
+void denormalize(void) {
+  if(fDenormalize) {
+    forv_Vec(FnSymbol, fn, gFnSymbols) {
+      denormalize(fn);
+    }
+  }
+}
+
+void denormalize(FnSymbol *fn) {
+
+  Vec<Symbol*> symSet;
+  Vec<SymExpr*> symExprs;
+  Map<Symbol*,Vec<SymExpr*>*> defMap;
+  Map<Symbol*,Vec<SymExpr*>*> useMap;
+
+  collectSymbolSetSymExprVec(fn, symSet, symExprs);
+  buildDefUseMaps(symSet, symExprs, defMap, useMap);
+
+  //data structures for deferred actual handling
+  std::set<CallExpr*> deferredFns;
+  ActualUseDefCastMap actualUseDefMap;
+
+  bool cachedGlobalManip = globalManipFuncCache.count(fn);
+
+  forv_Vec(Symbol, sym, symSet) {
+
+    SymExpr *use = NULL;
+    Expr *usePar = NULL;
+    Expr *def = NULL;
+    Expr *defPar;
+    Type* castTo = NULL;
+
+    //if we don't already have it cached,
+    //check for global symbols in function body
+    if(!cachedGlobalManip) {
+      //I am not sure if there should be !isConstant and !isParam as well
+      //I think there can be a const but stateful global that is manipulated by
+      //the function so, play safe
+      if(sym && !sym->isImmediate() && isGlobal(sym)){
+        globalManipFuncCache[fn] = true;
+        cachedGlobalManip = true;
+      }
+    }
+
+    if(isDenormalizable(sym, defMap, useMap, &use, &def, &castTo)) {
+      usePar = use->parentExpr;
+      defPar = def->parentExpr;
+
+      //defer if the symbol used as actual
+      if(CallExpr* useCe = toCallExpr(usePar)){
+        if(!useCe->isPrimitive()) {
+          deferredFns.insert(useCe);
+          std::pair<Expr*, Type*> defCastBundle(def, castTo);
+          actualUseDefMap.insert(std::pair<SymExpr*, std::pair<Expr*, Type*> >
+              (use, defCastBundle));
+          continue;
+        }
+      }
```

Comment:
I suspect this "defer actuals" business would not be necessary (and the code would be simpler) if you first decided, for each VarSymbol in a function's body, whether or not to denormalize it, and then, in a separate action, did the denormalizing.